providers/implementations/signature/rsa.c

```diff
@@ -0,0 +1,962 @@
+
+static int rsa_signature_init(void *vprsactx, void *vrsa)
+{
+    PROV_RSA_CTX *prsactx = (PROV_RSA_CTX *)vprsactx;
+
+    if (prsactx == NULL || vrsa == NULL || !RSA_up_ref(vrsa))
+        return 0;
+
+    RSA_free(prsactx->rsa);
+    prsactx->rsa = vrsa;
+    if (RSA_get0_pss_params(prsactx->rsa) != NULL)
+        prsactx->pad_mode = RSA_PKCS1_PSS_PADDING;
+    else
+        prsactx->pad_mode = RSA_PKCS1_PADDING;
+    /* Maximum for sign, auto for verify */
+    prsactx->saltlen = RSA_PSS_SALTLEN_AUTO;
+    prsactx->min_saltlen = -1;
+
+    return 1;
+}
+
+static int setup_md(PROV_RSA_CTX *ctx)
+{
+    if (ctx->mdname[0] != '\0') {
+        EVP_MD *md = EVP_MD_fetch(ctx->libctx, ctx->mdname, ctx->mdprops);
+
+        if (md == NULL)
+            return 0;
+        EVP_MD_CTX_free(ctx->mdctx);
+        EVP_MD_free(ctx->md);
+        ctx->md = md;
+        ctx->mdsize = EVP_MD_size(md);
+        ctx->mdctx = EVP_MD_CTX_new();
+        if (ctx->mdctx == NULL)
+            return 0;
+
+        if (!EVP_DigestInit_ex(ctx->mdctx, md, NULL))
+            return 0;
+    }
+
+    if (ctx->mgf1_mdname[0] != '\0') {
+        EVP_MD *md = EVP_MD_fetch(ctx->libctx, ctx->mgf1_mdname,
+                                  ctx->mgf1_mdprops);
+
+        if (md == NULL)
+            return 0;
+        EVP_MD_free(ctx->mgf1_md);
+        ctx->mgf1_md = md;
+    }
+
+    return 1;
+}
+
+static int setup_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        return 1;
+    if ((ctx->tbuf = OPENSSL_malloc(RSA_size(ctx->rsa))) == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    return 1;
+}
+
+static void clean_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        OPENSSL_cleanse(ctx->tbuf, RSA_size(ctx->rsa));
+}
+
+static void free_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        OPENSSL_clear_free(ctx->tbuf, RSA_size(ctx->rsa));
+}
+
+static int rsa_sign(void *vprsactx, unsigned char *sig, size_t *siglen,
+                    size_t sigsize, const unsigned char *tbs, size_t tbslen)
+{
+    PROV_RSA_CTX *prsactx = (PROV_RSA_CTX *)vprsactx;
+    int ret;
+
+    if (sig == NULL) {
+        *siglen = RSA_size(prsactx->rsa);
+        return 1;
+    }
+
+    if (sigsize < (size_t)RSA_size(prsactx->rsa))
+        return 0;
+
+    if (!setup_md(prsactx))
+        return 0;
+
+    if (prsactx->md != NULL) {
+        if (tbslen != prsactx->mdsize) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST_LENGTH);
+            return 0;
+        }
+
+        if (EVP_MD_type(prsactx->md) == NID_mdc2) {
+            unsigned int sltmp;
+
+            if (prsactx->pad_mode != RSA_PKCS1_PADDING) {
+                ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_PADDING_MODE,
+                               "only PKCS#1 padding supported with MDC2");
+                return 0;
+            }
+            ret = RSA_sign_ASN1_OCTET_STRING(0, tbs, tbslen, sig, &sltmp,
```

Comment:
NOOOOOOOOOOOOOOOO!!!!!!
All ASN1 needs to be outside the provider boundary.

Reply:
@davidmakepeace & @mattcaswell did some work in a PR related to doing special ASN1 code for the provider??

Reply:
PR #9138

Reply:
#9138 isn't merged yet, are you saying this PR should be pending on that?

Reply:
yes..

Reply:
Or we'll, it doesn't really have to wait. You might note that only the default provider is affected here, I haven't touched the FIPS module, and was actually imagining making that a separate PR, since as you noted, that requires a rewrite of the RSA library proper. So from *this* implementation, the calls should stay the same.

Reply:
PR #9138 can be done later.. All it does is a hardwired version of the ASN1 code..

Reply:
Right.

Reply:
BTW, a solution for this part is to have it appear in the legacy provider only...  considering MDC2 is only provided there anyway, that should be doable.

Reply:
Err in that case #ifndef FIPS_MODE will also do :)

Reply:
No